src/hotp/mod.rs

```diff
@@ -1,3 +1,85 @@
-pub fn generate() {}
+/// Default layer to generate a HOTP
+/// # Arguments
+///
+/// * `digest` - Byte digest generated by hashing the secret key
+///
+pub fn generate_default(
+    digest: Vec<u8>
+) -> String {
+    generate(None, digest)
+}
+
+
+/// Generates a HOTP
+///
+/// # Arguments
+///
+/// * `digits` - Optional parameter that defines the length of the one time pad (default: 6)
+/// * `digest` - Byte digest generated by hashing the secret key
+///
+pub fn generate(
+    digits: Option<u32>,
+    digest:  Vec<u8>, // Assume the digest is generated correctly. I don't like the cyclical nature that speakeasy has generateHotp -> digest -> generateHotp. Thoughts?
+) -> String {
+    let defined_digits = match digits {
+        Some(d) => d,
+        None => 6
+    };
+
+    let offset = match digest.last() {
+        Some(o) => o & 0xf,
+        None => 0 // 0 & 0xf == 0
+    };
+
+    /*
+    This section works to fill up the unsigned 32 bit number by:
+    1.  Taking the 8 bits at the offset from the digest, AND'ing with 0x7f so that we can ignore the sign bit
+        and then bit shifting 24 to the left to fill the most significant bits.
+    2.  Taking the next 8 bits from the digest at (offset + 1), AND'ing with 0xff to get the set bits, shifting 16 to fill
+        the next 8 significant bits.
+    3.  Same as (2.) but taking the bits from (offset + 2)
+    4.  Same as (2.) but taking the bits from (offset + 3)
+    5.  OR'ing each of these u32 so that we collapse all of the set bits into one u32
+     */
+    let f_msg = "Failed to pull from digest.";
+    let code = 
+        (u32::from(digest.get(offset as usize).expect(f_msg) & 0x7f) << 24) | 

```

Comment:
Lets not having any `expects` or `unwraps`.

We dont want the lib panicking.

Lets add an error to the GenerationErrors at the root crate and return that in a resullt